src/messaging/ExchangeContextDelegate.h

```diff
@@ -0,0 +1,120 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Exchange Context.
+ *
+ */
+
+#pragma once
+
+#include <system/SystemPacketBuffer.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+class ExchangeContext;
+
+/**
+ * @brief
+ *   This class provides a skeleton for the callback functions. The functions will be
+ *   called by ExchangeContext object on specific events. If the user of ExchangeContext
+ *   is interested in receiving these callbacks, they can specialize this class and handle
+ *   each trigger in their implementation of this class.
+ */
+class DLL_EXPORT ExchangeContextDelegate
```

Comment:
This needs some 'Free/Release/Dispose' method.

I do not know how CreateDelegate builds things... it could use heap, Platform::Memory*, object pools or static allocation.

When an exchange context is done, we need to clear it. It could be implicit in a timeout, however a exchange context cleanup seems to not map well to OnResponseTimeout.

Reply:
Agree. Since the delegate is bond to `ExchangeContext`, it is reasonable to add a `OnExchangeClosing` callback. Here comes my thought yesterday:

https://github.com/kghost/connectedhomeip/commit/84532d57e05b4b4ad4ef35a3682e9b99381f917b